### DISCUSSION TITLE
Real time pattern changes

### DISCUSSION THREAD

**Discussion by rodrigosge20** [pyswmm/pyswmm#515] [Q&A] [answered]
Hello!

I have a full year HOURLY pattern for the DWF in a .csv file (so 24*365 data points) which I would like to use as a Time Pattern in SWMM. The problem is SWMM itself only allows 1 day for HOURLY patterns (afaik), besides the MONTLY, DAILY and WEEKEND. Is it possible to overwrite this pattern through pyswmm? I also tried overwriting the inflow itself (Average value x Pattern), but I was not successful as well. 

What I am trying to do is to add a random pattern which changes every day.

Thank you!

**Comment by timothy-holmes** [comment]
Assuming you have a specific flow required at each hour, the SWMM solution
is to insert an INFLOW timeseries for the relevant period in the INP file,
generated using your csv file.

The pySWMM solution would be to load your inflow csv and use it to modify
the inflow as shown in the Ch 5.1 tutorial (
https://www.youtube.com/watch?v=av8L5gNSBvI).

**Reply by rodrigosge20** [reply]
Hey Timothy, thanks for the reply!

I understand your answer and I am actually using this function as well for another simulation. What I was more interested is if it is possible to change or overwrite a Time Pattern (or a single multiplier of the pattern) using pyswmm. For example, I have a HOURLY Time Pattern, which is a vector 24x1 with only ones (so [1 1 1 1 1 1 .... 1 1] for example), and I wanted after x simulation steps change one of the multipliers to 0.8, so this would imply a [0.8 1 1 1 1 1 ... 1 1] vector also 24x1. 

I thought on a workaround for that, but I was wondering if there is any way of doing something similar already through pyswmm. Thank you for the help!

**Reply by bemcdonnell** [reply]
@rodrigosge20 This is an interesting idea.  Changing the scaling factors along the way could be a pretty unique feature.  

However, I challenge the idea a little but because that is working within the standard bounds of how we used to think about SWMM.  I think @timothy-holmes's suggestion is quite good considering you would no longer be forced to think about a 24-hour set of peaking factors - now you have a 365*24 hour set of peaking factors * your average flow.  Something i'd be open to changing in SWMM is allowing the "node Inflow" timeseries to be the peaking factors but then allow the user to set another scaler value that can be multiplied by the peaking factor.  I need to noodle on that idea a little more - but i'd be open to it. 

Another consideration (which is way less optimal of an approach) is to make an INP for each day of the year and SAVE/USE a hotstart file to step the collection of simulations forward one day.  With each simulation you could modify the Hourly Time Patterns based on the measurements you have.   PySWMM-v2 has a new proposed feature that is awaiting approval (https://github.com/pyswmm/pyswmm/pull/512) that would allow you to adjust the parameters (not in the most elegant way).  

The first approach means that PySWMM is now simply part of the model, forever.  The latter approach leaves it SWMM-Friendly (but brings new challenges around the data management problem for 365 models).  You'd probably end up using Python tools to make use of the results anyways.  

How to pick!?! Good luck and keep us posted on your user experience!

**Reply by rodrigosge20** [reply, marked as answer]
Hello!

I was able to solve the problem with a non-optimal but ok solution. The way my code works, I run a parser of the .inp file before starting the simulation to be able to change it given specific parameters the user can set. So, within SWMM I created a pattern called ZERO, which is just a vector of 0s multipliers with size 24x1. Then, if I need this specific random pattern I created with 365x24 multipliers (hourly for a full year), I call my parser and change the .inp to start the simulation with all structures using this ZERO pattern, to be sure there is no DWF. During each simulation step I use the function generated_inflow() to inject a inflow using the 365*24 pattern I have (I read it from a .csv and create a df), making sure that I use the correct baseline (PE for example). The only thing to be careful is that generated_inflow() will accumulate the volume if I don't "remove" the last injected inflow. 

I hope I was clear in my explanation, is a little bit strange but for my purposes is working fine. By the way, thank you for the ongoing updates and follow-ups. I was really excited when the getters for pollution were added, you basically made my PhD 100% easier.
